osu.Game.Rulesets.Catch/Mods/CatchModHardRock.cs

```diff
@@ -1,13 +1,88 @@
 ﻿// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
 // Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
 
+using osu.Framework.MathUtils;
+using osu.Game.Rulesets.Catch.Objects;
 using osu.Game.Rulesets.Mods;
+using System;
 
 namespace osu.Game.Rulesets.Catch.Mods
 {
-    public class CatchModHardRock : ModHardRock
+    public class CatchModHardRock : ModHardRock, IApplicableToHitObject<CatchHitObject>
     {
         public override double ScoreMultiplier => 1.12;
         public override bool Ranked => true;
+
+        private float lastStartX;
+        private int lastStartTime;
+
+        public void ApplyToHitObject(CatchHitObject hitObject)
+        {
+            // Code from Stable, we keep calculation on a scale of 0 to 512
+            float position = hitObject.X * 512;
+            int startTime = (int)hitObject.StartTime;
+
+            if (lastStartX == 0)
+            {
+                lastStartX = position;
+                lastStartTime = startTime;
+                return;
+            }
+
+            float diff = lastStartX - position;
+            int timeDiff = startTime - lastStartTime;
+
+            if (timeDiff > 1000)
+            {
+                lastStartX = position;
+                lastStartTime = startTime;
+                return;
+            }
+
+            if (diff == 0)
+            {
+                bool right = RNG.NextBool();
+
+                float rand = Math.Min(20, (float)RNG.NextDouble(0, timeDiff / 4));

```

Comment:
Can we keep the positions relative to [0, 1] and only divide by `CatchPlayfield.BASE_SIZE` here?